### DISCUSSION TITLE
The JS app hooks ~will NOT~ might be available with the next release

### DISCUSSION THREAD

**Discussion by ganigeorgiev** [pocketbase/pocketbase#2825] [Announcements]
**UPDATE:** ~The JS app hooks will NOT be available with the next release.~ See the proposed changes in https://github.com/pocketbase/pocketbase/discussions/2825#discussioncomment-6376210.

---

Unfortunately it turned out that I spent the last ~3 weeks on the JS app hooks integration, documentation and testing for almost nothing because earlier today I realized one massive oversight on my part related to the js runtime (goja) and the fact that a single runtime instance can't work in a concurrent environment.

Although in our case we register the js hooks only once on application start, the hook callbacks usually operates as part of the requests flow, and since we handle each request in its own goroutine, the hooks could panic as they will not have safe concurrent access to the goja runtime (see https://github.com/dop251/goja/issues/522).
One possible workaround I guess could be to have a wrapper function that will create a new js runtime for each hook call, but the problem with this approach is that first the performance will significantly degrade as this will be executed on each hook call and second - it will require heavily modifying the goja internals since we currently don't have direct access to the local declarations (let, const, var, etc.) to copy from one vm to another.

At the end, I've decided to give up this feature and to not spent anymore time on it since it has a very high maintenance cost and it is better to focus on something more productive.

To be fair, it wasn't a complete waste of time because with the current integration we finally have:
- documentation and automated TypeScript ambient declarations for the JS migrations
- new more detailed Go guides, although tomorrow I'll have to refactor some of the layout changes since currently it looks like in the screenshot below and will have to remove the JS parts that don't make sense outside of a migration context:
![docs_failure](https://i.imgur.com/f4r2ut6.png)

---

**Anyway, sorry for the false JS hooks promise and the release freeze.**

I'll try to cleanup the mess within the next 1-2 weeks and will prepare a v0.17.0 release.

_The current jsvm plugin integration and site documentation will be available under the `js_hooks_failure` branches in case sometime in the future we decide to get back to it._

**Comment by benallfree** [comment]
@ganigeorgiev You made it a lot farther than I did! So maybe I’ll resume developing Deno-based cloud workers for https://pockethost.io.

**Comment by MyWay** [comment]
Thank you for your effort.

**Comment by dschissler** [comment]
@ganigeorgiev Perhaps you could reach out to other scripting language systems to see if there is something else that could be worked in.  It might yield some ideas or a competing paradigm somehow.  Maybe a Javascript-like syntax, but more custom built for this sort of thing.  Also maybe WASM?  Maybe there are ways of compiling dynamically, etc.  I mean that maybe you should take a deep breath and to look at the entire landscape while this is still in your head.  Zoom out to how something else similar could work, or if it will just never work at all.  If you drop it all now then you'll be at risk of forgetting some of the nitty gritty details.  I find that unless I'm under a strict deadline that it's a mistake to completely abandon a failed research project without dreaming a little bit.

* [Waxix](https://wasix.org/), [Wasmer 4.0 WebAssembly Implementation Brings WASIX Support](https://www.phoronix.com/news/Wasmer-4.0)
* [Wazero](https://wazero.io/), [Wazero: Zero dependency WebAssembly runtime written in Go](https://news.ycombinator.com/item?id=36551617)
* Perhaps something like this but for Golang [Javy](https://github.com/bytecodealliance/javy)
* [Tengo -  A fast script language for Go ](https://github.com/d5/tengo)
* [Using WebAssembly to turn Rust crates into fast TypeScript libraries](https://rybicki.io/blog/2023/06/27/rust-crate-into-typescript-library.html)

[edit] What about writing it in Javascript, Pocketbase converts it to WASM and then runs it?  Pocketbase could put up a temporary message that maintenance is being performed, etc while it's compiled.

**Reply by ganigeorgiev** [reply]
WASM integration was suggested before (see https://github.com/pocketbase/pocketbase/discussions/410) but for now remains out of the scope as it would require maintaining 2 different apis. The whole point of a interpreter language like goja was to allow easily sharing go struct and methods without having to create manually bindings and documentations for everything (see also [pocketbase/tygoja](https://github.com/pocketbase/tygoja)).
There are some workarounds for the above mentioned issue, including forking and modifying the goja internals to adapt the PocketBase specific use case but it would be too hard to maintain in the long run and at least for now I don't feel that investing more time on this will be beneficial as there are plenty of other blocker tasks to work on (transactions, logs generalization, realtime filter and expand support, api rules improvements, go autogenerated collection models, etc.).

**Comment by Lex-2008** [comment]
Thanks for the announcement! I wonder if it might be possible to expose a separate API endpoint which, when accessed, would execute some goja JS function? It can probably be serialized to avoid working on several requests in parallel. In the worst case, it can return some "503 goja backend is busy" error and let clients implement a retry-after-delay strategy.

On the other hand, this is not much different from a separate deno/nodejs app running next to PocketBase and making api requests.

**Reply by ganigeorgiev** [reply]
> if it might be possible to expose a separate API endpoint which, when accessed, would execute some goja JS function?

It is possible, but the main idea behind the JS app hooks was to allow low level app interactions in a similar manner to the existin Go hooks, eg. intercepting record create requests, executing custom db queries, registering custom routes, custom console commands, cron-like tasks scheduling, etc.

I've started working on a testbed and after its completion I'll give the JS app hooks one last chance to test whether creating a new goja runtime on each hook or request call will hurt too much the performance.

**Reply by dschissler** [reply]
@ganigeorgiev Perhaps even if it's super slow that you could still leave it in there but with heavy warnings that it's not performant at all.  It could still be quite good for things like home automation, a website for some contractor who only gets 0.001req/s.  Then you could come back to it.  Unless it's just slow motion snail speed then what is wrong with saying that it's **SLOW**, but that someday it might not be?  Also maybe once it's in there some wizard will need it and will swoop in to save the day.

**Reply by ganigeorgiev** [reply]
I'll try to salvage whatever I can from the current implementation, but no hard promises.

**Comment by RepComm** [comment]
I didn't know this was being worked on and I can't tell you how excited I am! This project is fire.

**Comment by 171906502** [comment]
how is lua? it has not performance problem

**Reply by ganigeorgiev** [reply]
Lua itself is not that well known as a language, especially by mobile and web developers.
Additionally it will require creating and maintaining manual bindings for almost everything (see https://github.com/pocketbase/pocketbase/pull/1729) and the main reason to chose goja was that it will do the heavy lifting for us since we map only directly some entrypoint go variables and functions.

The raw performance for example between goja and go-lua would be roughly the same when calling wrapped go functions like in our case and replacing the vm with another interpreter unfortunately will not solve the concurrency issue (_it was just an oversight on my part and I shouldn't expected the callback functions and the runtime stack to magically operate properly in a concurrent environment and you'll need some sort of locking/synchronization mechanism to do that safely, but this technically means that all registered interpreter hooks and routes will have to be processed sequential diminishing the benefit of having a go backend in the first place_).

**Reply by 171906502** [reply]
If it's just a performance issue, a goja  pool may alleviate it.The benefits of supporting dynamic scripting are more cost-effective than performance damage.

**Comment by tigawanna** [comment]
I stumbled on  this]([https://dev.to/jacktt/plugin-in-golang-4m67?utm_source=dormosheio&utm_campaign=dormosheio), I don't know how feasible it would be for this purpose , if JavaScript fails a plain go one that's dynamically loaded wouldn't be that bad too [

**Reply by ganigeorgiev** [reply]
Unfortunately the current native Go plugins have several limitations:

- they are no portable
- require both the main code base and the plugin to be compiled with the exact Go version
- no Windows support

_(see also [`plugins` Warning](https://pkg.go.dev/plugin#hdr-Warnings))_

**Reply by tigawanna** [reply]
Those are definitely deal breakers

**Comment by whmeitzler** [comment]
Does Goja's [RunOnLoop](https://github.com/dop251/goja_nodejs/blob/804a84515562332a36e8301c37c779fafe163f75/eventloop/eventloop.go#L248) function not help? If you want to maintain the context of the app itself, rather than the lifetime of a single request, that may be the best you can do. And, for the record, it looks pretty damn good. NodeJS devs already accept single threaded paradigms (whether they know/like it or not), and Pocketbase shipping with all its batteries included, *plus no node_modules* has a lot of value. You've done good work here, it'd be a shame to delete it outright due to a ceiling in perf. 

If you want to add parallelism, you could split the runtime into the concept of `pb_service_hooks` and `pb_request_hooks`. Each have a different lifetime and operational scope, with `pb_service_hooks`operating at the lifetime of the app itself and running on a proper event loop, and `pb_request_hooks` being just the Goja VM that can be executed sequentially within the context of a single PB request. 

Communication between these instances can be achieved by `pb_request_hooks` triggering...hooks... that `pb_service_hooks` then react to, similar to channels in Go. Or you can simply offer a few examples of porting this kind of work from JS to Go and kindly tell people that, if they're hitting perf limits, it may be better to port to Go.

**Reply by ganigeorgiev** [reply]
I don't see how `RunOnLoop` will help in our case (_and in fact we are already using it_).

The major issue is not just the performance, but the fact that **it requires creating, documenting and maintaining a different set of APIs**, which defeat the purpose of why I decided to integrate goja in the first place. For example, you no longer can have `$app.onModelBeforeCreate().add(callback)` by just binding `vm.Set("$app", app)`. Goja also doesn't expose many of its internals so there is no easy way without forking to automate some of the bindings by copying the existing runtime stack and its local let, const, function declarations.

Yes, there are existing workarounds, we can create wrapper functions, maintain a pool of vms, etc. but similar to the [Lua PR](https://github.com/pocketbase/pocketbase/pull/1729), I'm not sure I want to go that route as it comes with a very high maintenance cost. As mentioned earlier, I'll try to salvage whatever I can in the following days, but I'm not confident that the results will be much different.

**Comment by ganigeorgiev** [comment]
I gave it a little more thought after some of the above comments, and I think that actually it may be possible to preserve most of the current implementation without having to create manual bindings for everything and still maintain the same near identical Go performance.

The application will run 2 types of goja runtimes:
- **Loader** (_created once on startup_)
- **Executors** (created inside each fired goroutine)

On startup the Loader runtime will kick in and will read all the `.pb.js` files.
This Loader runtime will have 4 special types of top-level functions available (_I'm not sure about the namings yet_):
 1) `on*(handler, [...optTags])` - for each all app level hooks with exception of the `OnBeforeServe`. This bindings and their documentation can be automated with reflection (the slight overhead is not important here as it will happen only once during the loading phase)
 2) `cronAdd(name, expr, handler)`, `cronRemove(name)` - for managing cron jobs that will execute periodically at the specified cron expression time
 3) `routerAdd(method, path, handler, ...middlewares)` - for registering a new echo route handler
 4) `routerPre(...middlewares)`, `routerUse(...middlewares)` -  for registering global middlewares (aka. equivalent to `echo.Router.Pre()` and `echo.Router.Use()`).


1), 2) and 3) are similar in terms of how the `handler` argument would be treated. Once executed they will:
- serialize the JS `handler` argument as a plain string
- pass the serialized function to `goja.Compile()` to generate a precompiled "blank program" that will be reused across different goroutines
- then we register the equivalent Go wrapping handler (either via reflection or directly) that will create a *new Executor runtime every time when it is called in a goroutine with the precompiled program (_*could use a buffer/pool_)

The `middlewares` argument in 3) and 4) are slightly more different because they are variadic and mixed since we also have "native" middlewares exposed (eg. `$apis.requireAdminAuth`), meaning that we can't just serialize every passed argument as string, so an additional type conversion will have to be performed behind the scenes to determine whether it needs to be treated as a goja script or run directly as `echo.MiddlewareFunc` (_this should be safe for concurrent use_).

In all cases the precompiled program will be generated with something like:
```js
p := goja.MustCompile("", "("+ handlerAsString +").apply(undefined, args)", true)

// for the middlewares, since they return an additional handler function this could look something like this:
p := goja.MustCompile("", "(("+middlewareAsString+").apply(undefined, args)).apply(undefined, args2)", true)
```

`args` will be passed from each executor by doing something like:
```go
// initialize the executor...
executor.Set("args", []any{"something"})
_, err := executor.RunProgram(p)
executor.Set("args", goja.Undefined())
// release the executor...
```

When combined with a buffer/pool (eg. 200 goja runtimes), the below Go and JS samples have similar execution times with 100 concurrent clients:

```go
// Go:
customMiddleware := func(next echo.HandlerFunc) echo.HandlerFunc {
  return func(c echo.Context) error {
    c.Set("total", 20)

    return next(c)
  }
}

app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
  e.Router.GET("/go", func(c echo.Context) error {
    total := c.Get("total").(int)

    records, err := app.Dao().FindRecordsByFilter("example", "title != ''", "-created", total)
    if err != nil {
      return err
    }

    return c.JSON(200, records)
  }, apis.ActivityLogger(app), customMiddleware)

  return nil
})
```

```js
// JS:
function customMiddleware(next) {
    return (c) => {
        c.set("total", 20)

        next(c)
    }
}

routerAdd("GET", "/js", (c) => {
    const total = c.get("total")

    const records = $app.dao().findRecordsByFilter("example", "title != ''", "-created", total)

    c.json(200, records)
}, $apis.activityLogger($app), customMiddleware);
```

**As you've probably already noticed, the main drawback with the functions serialization approach is that you can't reuse variables and functions declared in the top-level scope, aka. the following will fail:**

```js
var secret = "something"

onModelBeforeCreate((e) => {
   console.log(secret) // <-- this will be undefined
})
```

One slightly less-convenient workaround for sharing/reusing code parts, could be to move the reusable code in a separate helper `.js` file and import it with `require()` inside the handler:

```js
// config.js
module.exports = {
    "secret": "something"
}

// main.pb.js
onModelBeforeCreate((e) => {
   const config = require("/path/to/pb_hooks/config.js");

   console.log(config.secret)
})
```

Everything else (dao, db query builder, sending emails, etc.) will work as it is right now without further extra bindings.

Internally it will be a little messy, but hopefully it would be something that rarely will have to change.

Unless again I'm ignoring something obvious, I think the above approach, even considering its limitations, should work fine for most cases. If I'm missing something or someone has better idea, please free to let me know and I'll consider it while cleaning up the implementation.

**Comment by whmeitzler** [comment]
I like where this is going; thanks for pushing so hard for autocomplete/Typescript bindings. 
I would suggest avoiding `require()` as much as possible; it can further confuse when/where code is running. 
If loaders want to pass configuration into executors, there is already a robust database available for sharing state. I would suggest that, or, if perf is a concern in the future, a Redis-style kv store baked in, rather than trying to force state into your eval() at just the right time.

**Reply by ganigeorgiev** [reply]
`require()` will be something optional. I'll make sure to document it and outline its caveats.
